File: python/protein-translation/protein_translation.py
from amino_acids import AminoAcid as Amino

U, C, A, G = "U", "C", "A", "G"  # Avoid excessive typing.
STOP = "STOP"

RNA_CODON_TREE = {
    U: {
        U: {U: Amino.Phe, C: Amino.Phe, A: Amino.Leu, G: Amino.Leu,},
        C: Amino.Ser,
        A: {U: Amino.Tyr, C: Amino.Tyr, A: STOP, G: STOP,},
        G: {U: Amino.Cys, C: Amino.Cys, A: STOP, G: Amino.Trp,},
    },
    C: {
        U: Amino.Leu,
        C: Amino.Pro,
        A: {U: Amino.His, C: Amino.His, A: Amino.Gln, G: Amino.Gln,},
        G: Amino.Arg,
    },
    A: {
        U: {U: Amino.Ile, C: Amino.Ile, A: Amino.Ile, G: Amino.Met,},
        C: Amino.Thr,
        A: {U: Amino.Asn, C: Amino.Asn, A: Amino.Lys, G: Amino.Lys,},
        G: {U: Amino.Ser, C: Amino.Ser, A: Amino.Arg, G: Amino.Arg,},
    },
    G: {
        U: Amino.Val,
        C: Amino.Ala,
        A: {U: Amino.Asp, C: Amino.Asp, A: Amino.Glu, G: Amino.Glu,},
        G: Amino.Gly,
    },
}


def express_codon(codon):
    tree = RNA_CODON_TREE[codon[0]][codon[1]]
    return tree if isinstance(tree, Amino) else tree[codon[2]]


def proteins(strand):
    protein = []
    for triplet in [strand[i : i + 3] for i in range(0, len(strand), 3)]:
        amino = express_codon(triplet)
        if amino == STOP:
            break
        protein.append(amino.value)
    return protein

File: javascript/protein-translation/protein-translation.js
const AMINO_ACIDS = {
  Ala: "Alanine",
  Arg: "Arginine",
  Asn: "Asparagine",
  Asp: "Aspartic",
  Cys: "Cysteine",
  Gln: "Glutamine",
  Glu: "Glutamic",
  Gly: "Glycine",
  His: "Histidine",
  Ile: "Isoleucine",
  Leu: "Leucine",
  Lys: "Lysine",
  Met: "Methionine",
  Phe: "Phenylalanine",
  Pro: "Proline",
  Ser: "Serine",
  Thr: "Threonine",
  Trp: "Tryptophan",
  Tyr: "Tyrosine",
  Val: "Valine"
};

const RNA_CODON_TREE = {
  U: {
    U: { U: "Phe", C: "Phe", A: "Leu", G: "Leu" },
    C: { U: "Ser", C: "Ser", A: "Ser", G: "Ser" },
    A: { U: "Tyr", C: "Tyr", A: "STOP", G: "STOP" },
    G: { U: "Cys", C: "Cys", A: "STOP", G: "Trp" }
  },
  C: {
    U: { U: "Leu", C: "Leu", A: "Leu", G: "Leu" },
    C: { U: "Pro", C: "Pro", A: "Pro", G: "Pro" },
    A: { U: "His", C: "His", A: "Gln", G: "Gln" },
    G: { U: "Arg", C: "Arg", A: "Arg", G: "Arg" }
  },
  A: {
    U: { U: "Ile", C: "Ile", A: "Ile", G: "Met" },
    C: { U: "Thr", C: "Thr", A: "Thr", G: "Thr" },
    A: { U: "Asn", C: "Asn", A: "Lys", G: "Lys" },
    G: { U: "Ser", C: "Ser", A: "Arg", G: "Arg" }
  },
  G: {
    U: { U: "Val", C: "Val", A: "Val", G: "Val" },
    C: { U: "Ala", C: "Ala", A: "Ala", G: "Ala" },
    A: { U: "Asp", C: "Asp", A: "Glu", G: "Glu" },
    G: { U: "Gly", C: "Gly", A: "Gly", G: "Gly" }
  }
};

const invalid = seq => seq && seq.match(/[^ACGU]/);

const triplets = seq => (seq && seq.match(/[ACGU]{3}/g)) || [];

const decode = codon => [...codon].reduce((a, b) => a[b], RNA_CODON_TREE);

export const translate = seq => {
  if (invalid(seq)) {
    throw new Error("Invalid codon");
  }
  let aminos = triplets(seq).map(codon => decode(codon));
  const stop = aminos.findIndex(e => e === "STOP");
  if (stop >= 0) {
    aminos = aminos.slice(0, stop);
  }
  return aminos.map(a => AMINO_ACIDS[a]);
};
